File: py/37.py
from math import sqrt

def prime(n):
    if n < 3:
        return n - 1

    if n % 2 == 0:
        return False

    for i in xrange(3, int(sqrt(n) + 1), 2):
        if n % i == 0:
            return False

    return True

def truncatable(n):
    if not prime(n):
        return False

    dup, mag = n, 10
    while dup:
        dup /= 10
        if not prime(dup):
            return False

    dup = n
    while dup:
        dup /= 10
        mag *= 10

    while n:
        mag /= 10
        if not prime(n):
            return False
        n -= n / mag * mag

    return True

print reduce(lambda a, b: a + b,
        filter(truncatable, xrange(11, 750000)))

File: c/37.c
#include <stdio.h>
#include <math.h>

int prime(int n)
{
	int i;

	if (n < 3)
		return n - 1;

	if (n % 2 == 0)
		return 0;

	for (i = 3; i <= sqrt(n); i += 2)
		if (n % i == 0)
			return 0;

	return 1;
}

int truncatable(int n)
{
	int dup = n, mag = 10;

	if (!prime(n))
		return 0;

	while (dup /= 10)
		if (!prime(dup))
			return 0;

	dup = n;
	while (dup /= 10) mag *= 10;

	while (n) {
		mag /= 10;
		if (!prime(n))
			return 0;
		n -= n / mag * mag;
	}

	return 1;
}

int main(void)
{
	int i, sum = 0;

	for (i = 11; i < 750000; ++i)
		sum += truncatable(i) ? i : 0;

	printf("%d\n", sum);
	return 0;
}
